### PR TITLE
feat(iroh-relay)!: use explicit key cache

### PR DESCRIPTION
**Pull request by rklaehn**
## Description

This wires up an explicit key cache to replace the implicit one that was removed in #3051.

~~The default for a key cache is Disabled. A disabled key cache has a size of 1 pointer and otherwise zero performance overhead.~~

I have removed the Default instance for both KeyCache and DerpProtocol so you don't accidentally pass the default despite having a cache available.

We use the lru crate for the cache for now. Please comment if it should be something else.

Benchmarks have shown that conversion from a [u8;32] to a VerifyingKey is relatively cheap, so the purpose of the cache is solely to validate incoming public keys.

We add a Borrow instance to PublicKey so we can use it as a cache key.

Some performance measurements:

```
Benchmarking validate valid ed25519 key: 3.7674 µs
Benchmarking validate invalid ed25519 key: 3.6637 µs
Benchmarking validate valid iroh ed25519 key: 67.089 ns
Benchmarking validate invalid iroh ed25519 key: 64.004 ns
```

So just from validating incoming keys, without cache you would be limited to ~250 000 msgs/s per thread. At a message size of 1KiB that would be 250MB/s, which is not great.

With the cache deserialization can do 14 000 000 msgs/s, which means that this is no longer a bottleneck.

## Breaking Changes

- RelayConfig has new public field key_cache_capacity

## Notes & open questions

- Size of the cache
- source_and_box has a PublicKey::try_from. Is that perf critical?

## Change checklist

- [ ] Self-review.
- [ ] Documentation updates following the [style guide](https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.html#appendix-a-full-conventions-text), if relevant.
- [ ] Tests if relevant.
- [ ] All breaking changes documented.

### CODE REVIEW

File: iroh-relay/Cargo.toml [hunk 1]
```diff
@@ -96,6 +96,7 @@ url = { version = "2.5", features = ["serde"] }
 webpki = { package = "rustls-webpki", version = "0.102" }
 webpki-roots = "0.26"
 data-encoding = "2.6.0"
+lru = "0.12.5"
```
Comment (flub) [inline iroh-relay/Cargo.toml:99]: Purely a drive-by comment.  But I've had another reason to add an LRU cache to iroh-relay very recently.  So I think this dependency would make it here at some point.

**Comment by dignifiedquire** [comment]
I think a default of maybe 128 or so would be good, as it is very small still but will improve perf for most basic endpoint usages

**Comment by rklaehn** [comment]
> I think a default of maybe 128 or so would be good, as it is very small still but will improve perf for most basic endpoint usages

I made 128 the default for client and 1024 * 1024 the default for a http server. Presumably for the server 32 megabytes (or 56 megabytes if you add the 2 ptrs that LruEntry adds and the key) is no big deal, and it will work up to 1m connections.

File: iroh-relay/src/server/http_server.rs [hunk 2]
```diff
@@ -172,6 +175,7 @@ impl ServerBuilder {
             handlers: Default::default(),
             headers: HeaderMap::new(),
             client_rx_ratelimit: None,
+            key_cache_capacity: 1024 * 1024,
```
Comment (dignifiedquire) [inline iroh-relay/src/server/http_server.rs:178]: maybe add a comment why this number was choosen

File: iroh-relay/src/key_cache.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    Disabled,
+    /// The key cache is enabled with a fixed capacity. It is shared between
+    /// multiple threads.
+    Shared(Arc<Mutex<lru::LruCache<PublicKey, ()>>>),
+}
+
+impl KeyCache {
+    /// Key cache to be used in tests.
+    #[cfg(test)]
+    pub fn test() -> Self {
```
Comment (dignifiedquire) [inline iroh-relay/src/key_cache.rs:31]: this is the same as `KeyCache::default()`, why is this needed?

File: iroh-relay/src/key_cache.rs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    /// Get a key from a slice of bytes.
+    pub fn key_from_slice(&self, slice: &[u8]) -> Result<PublicKey, SignatureError> {
+        let Self::Shared(cache) = self else {
+            return PublicKey::try_from(slice);
+        };
+        let Ok(bytes) = PublicKeyBytes::try_from(slice) else {
+            // if the size is wrong, use PublicKey::try_from to fail with a
+            // SignatureError.
+            PublicKey::try_from(slice)?;
+            unreachable!();
```
[the remark this answers is not in the document]
Comment (rklaehn) [reply iroh-relay/src/key_cache.rs:55]: PublicKey::try_from(slice)?; is just a fancy way to return an error. It is guaranteed to fail when slice is not 32 bytes long.

File: iroh-relay/src/key_cache.rs (hunk 3 above)
Comment (rklaehn) [reply iroh-relay/src/key_cache.rs:31]: I was looking through the code for places where KeyCache::default is used to make sure I got them all. So I renamed this to test so it is clear that it is only for tests.

File: iroh-relay/src/key_cache.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply iroh-relay/src/key_cache.rs:55]: this reads a bit nicer
```rust
return Err(PublicKey::try_from(slice).expect_err("invalid length"));
```

File: iroh-relay/src/key_cache.rs (hunk 3 above)
Comment (rklaehn) [reply iroh-relay/src/key_cache.rs:31]: I have now done the same for DerpCodec and removed the Default instance. I think with Default it is too easy to accidentally have a place where the cache is not used at all.

File: iroh-relay/src/protos/relay.rs [hunk 5]
```diff
@@ -204,8 +206,23 @@ pub(crate) async fn recv_client_key<S: Stream<Item = anyhow::Result<Frame>> + Un
...
-#[derive(Debug, Default, Clone)]
-pub(crate) struct DerpCodec;
+#[derive(Debug, Clone)]
+pub(crate) struct DerpCodec {
+    cache: KeyCache,
+}
+
+impl DerpCodec {
+    #[cfg(test)]
+    pub fn test() -> Self {
```
Comment (dignifiedquire) [inline iroh-relay/src/protos/relay.rs:216]: `pub(crate)`?

**Comment by dignifiedquire** [comment]
I am having a hard time finding the place this is configured for a non server endpoint, can you point me to that?

**Comment by rklaehn** [comment]
> I am having a hard time finding the place this is configured for a non server endpoint, can you point me to that?

It is configurable using the ClientBuilder, with a default of 128 as you suggested.

```rust
  /// Create a new [`ClientBuilder`]
  pub fn new(url: impl Into<RelayUrl>) -> Self {
      ClientBuilder {
          is_preferred: false,
          address_family_selector: None,
          is_prober: false,
          server_public_key: None,
          url: url.into(),
          protocol: Protocol::Relay,
          #[cfg(any(test, feature = "test-utils"))]
          insecure_skip_cert_verify: false,
          proxy_url: None,
          key_cache_capacity: 128,
      }
  }
...


    /// Set the capacity of the cache for public keys.
    pub fn key_cache_capacity(mut self, capacity: usize) -> Self {
        self.key_cache_capacity = capacity;
        self
    }
  ```

**Review by dignifiedquire — APPROVED** [review]

**Comment by dignifiedquire** [comment]
> Is that not the right place?

It is, I was just being blind
